sidecar/tlv.go

```diff
@@ -0,0 +1,377 @@
+package sidecar
+
+import (
+	"io"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightninglabs/pool/order"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+const (
+	idType        tlv.Type = 1
+	versionType   tlv.Type = 3
+	stateType     tlv.Type = 5
+	offerType     tlv.Type = 7
+	recipientType tlv.Type = 9
+	orderType     tlv.Type = 11
+	executionType tlv.Type = 13
+
+	offerLength     = 8 + 8 + 33 + 64
+	recipientLength = 33 + 33
+	orderLength     = 32 + 33 + 64
+	executionLength = 32
+)
+
+var (
+	// ZeroPendingChannelID is an empty pending channel ID with all bits set
+	// to zero.
+	ZeroPendingChannelID = [32]byte{}
+
+	// ZeroPubKey is an empty public key with all bits set to zero.
+	ZeroPubKey = [33]byte{}
+
+	// ZeroSignature is an empty signature with all bits set to zero.
+	ZeroSignature = [64]byte{}
+)
+
+// SerializeTicket binary serializes the given ticket to the writer using the
+// tlv format.
+func SerializeTicket(w io.Writer, ticket *Ticket) error {
+	version := uint8(ticket.Version)
+	state := uint8(ticket.State)
+
+	records := []tlv.Record{
+		tlv.MakeStaticRecord(idType, &ticket.ID, 8, EBytes8, DBytes8),
+		tlv.MakePrimitiveRecord(versionType, &version),
+		tlv.MakePrimitiveRecord(stateType, &state),
+		tlv.MakeStaticRecord(
+			offerType, &ticket.Offer, offerLength, EOffer, DOffer,
+		),
+	}
+
+	if ticket.Receiver != nil {
+		records = append(records, tlv.MakeStaticRecord(
+			recipientType, &ticket.Receiver, recipientLength,
+			ERecipient, DRecipient,
+		))
+	}
+	if ticket.Order != nil {
+		records = append(records, tlv.MakeStaticRecord(
+			orderType, &ticket.Order, orderLength, EOrder, DOrder,
+		))
+	}
+	if ticket.Execution != nil {
+		records = append(records, tlv.MakeStaticRecord(
+			executionType, &ticket.Execution, executionLength,
+			EExecution, DExecution,
+		))
+	}
+
+	tlvStream, err := tlv.NewStream(records...)
```

Comment:
I thought the idea was to use _two_ logical streams to allow their contents to be signed independently?

Reply:
Discussed this offline. We did mention using two streams initially but during implementation I noticed that the pattern of having a distinct `Digest()` method and then embedding the signature inside the TLV stream itself makes things easier. Mostly for also storing a ticket in the DB as we can just serialize the whole stream and put it into `etcd` or `bbolt`.